Add a config option to enable Flyway baseline feature

rmanibus opened the pull request:
Resolves #27734

rmanibus committed:
flyway: config option to baseline new environment at startup

geoand commented:
Thanks for this.

It seems like the test does not compile

rmanibus commented:
@geoand I was struggling to get the project imported correctly by intellij, it should be good now.

I also rephrased the baselineOnMigrate config doc because the current description is a bit misleading

rmanibus commented:
@geoand weird that FlywayExtensionMigrateAtStartTest is failing. H2 is not reset between tests ?

geoand commented:
I'm pretty sure it's not

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:
```diff
@@ -69,6 +70,10 @@ public Flyway get() {
 
     public void doStartActions() {
         for (FlywayContainer flywayContainer : FLYWAY_CONTAINERS) {
+            if (flywayContainer.isBaselineAtStart() &&
+                    Objects.isNull(flywayContainer.getFlyway().info().current())) {
```
```suggestion
            if (flywayContainer.isBaselineAtStart() &&
                    flywayContainer.getFlyway().info().current() == null) {
```

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:
```diff
@@ -69,6 +70,10 @@ public Flyway get() {
 
     public void doStartActions() {
         for (FlywayContainer flywayContainer : FLYWAY_CONTAINERS) {
+            if (flywayContainer.isBaselineAtStart() &&
+                    Objects.isNull(flywayContainer.getFlyway().info().current())) {
+                flywayContainer.getFlyway().baseline();
+            }
```
I think we should analyze the order on how this change affects the others `*AtStart` flags. 
For example: 
- Do we have any mutual exclusion? 
- Should `clean()` be always executed first?

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:
```diff
@@ -113,11 +113,20 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
     public boolean baselineOnMigrate;
 
+    /**
+     * true to execute Flyway baseline automatically when the application starts when executed against an empty schema, false
```
Apparently this command fails if the `flyway_schema_history` table exists (see https://github.com/quarkusio/quarkus/issues/27734#issuecomment-1238813838).

rmanibus replied:
I will add a test with an existing table to be sure

rmanibus replied:
@gastaldi , I made some test and here is my conclusion:

if the flyway_schema_history table exist and contains any element, everything is ok.
if the flyway_schema_history does not exist, everything is ok.
if the flyway_schema_history exist but is empty, then flyway.info().current() is null so we do execute the baseline and it is failing. 

I will try to find a way to check if flyway_schema_history already exist in flyway api.

gastaldi commented:
> @geoand weird that FlywayExtensionMigrateAtStartTest is failing. H2 is not reset between tests ?

Take a look at how it is done in https://github.com/quarkusio/quarkus/blob/main/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionCallbackTest.java

rmanibus commented:
> For the record, I'm running full Quarkus builds locally (completely unrelated to this PR) and I'm seeing `FlywayExtensionMigrateAtStartTest` fail as well. So I believe that's not caused by this PR, but sneaked in through some other PR.

@Ladicek I just changed the DB name in the connection string of my test and everything is working fine now

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:
```diff
@@ -72,6 +73,15 @@ public void doStartActions() {
             }
+            if (flywayContainer.isBaselineAtStart()) {
+                new FlywayExecutor(flywayContainer.getFlyway().getConfiguration())
+                        .execute((m, schemaHistory, d, ds, s, ce, si) -> {
+                            if (!schemaHistory.exists()) {
+                                return flywayContainer.getFlyway().baseline();
+                            }
+                            return null;
+                        }, true);
```
Please avoid lambdas in the recorder code

gsmet commented:
@gastaldi can you have another look?

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:
```diff
@@ -113,11 +113,20 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
     public boolean baselineOnMigrate;
 
+    /**
+     * true to execute Flyway baseline automatically when the application starts when executed against an empty schema, false
```
```suggestion
     * true to execute Flyway baseline automatically when the application starts. This flag is ignored if the flyway_schema_history table does not exist in the current schema.
```

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:
```diff
@@ -113,11 +113,20 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
 
     /**
-     * Enable the creation of the history table if it does not exist already.
+     * call baseline when migrate is executed against a non-empty schema with no metadata table.
+     * This schema will then be baselined with the baselineVersion before executing the migrations.
+     * Only migrations above baselineVersion will then be applied
```
This is confusing, can you reword it?

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:
```diff
@@ -80,4 +92,21 @@ public void doStartActions() {
+        @Override
+        public BaselineResult execute(CompositeMigrationResolver cmr, SchemaHistory schemaHistory, Database d,
+                Schema defaultSchema, Schema[] s, CallbackExecutor ce, StatementInterceptor si) {
+            if (!schemaHistory.exists()) {
```
You said previously:
> if the flyway_schema_history table exist and contains any element, everything is ok.

Why just checking if the `flyway_schema_history` does not exist here?

gastaldi reviewed (changes requested):
I added some comments. I think it still needs some clarification

geoand commented:
This is going to need a rebase

rmanibus commented:
@geoand I rebased it

geoand commented:
Thanks!

Have the comments made during the review been addressed? cc @gastaldi

rmanibus commented:
If I remember well I addressed all the concerns.
I could also include a guide on how to perform migrations squashing in the documentation.

gastaldi commented on extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:
```diff
@@ -141,11 +141,22 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
 
+    /**
+     * true to execute Flyway baseline automatically when the application starts.
+     * This flag is ignored if the flyway_schema_history table exist in the current schema.
```
```suggestion
     * This flag is ignored if the flyway_schema_history table exists in the current schema.
```

gastaldi reviewed (approved):
LGTM, one minor observation

rmanibus commented:
> If I remember well I addressed all the concerns. I could also include a guide on how to perform migrations squashing in the documentation.

fixed the typo. I will do this in another PR

gastaldi commented:
Could you squash the commits before we trigger CI?

Sanne commented:
@gastaldi ping ? This looks like ready but it's waiting for your opinion

gastaldi commented:
@Sanne thanks, I've rebased the branch, it should be good to merge once it passes

gastaldi commented:
@rmanibus because I forgot to merge this PR 3 months ago, this now needs to be changed to match the latest Flyway codebase. Can you have a look?

rmanibus commented:
@gastaldi it is rebased !

gastaldi commented:
Can you rebase and remove the merge commit?

rmanibus commented:
@gastaldi done !
